engineering/conventions.md

```diff
@@ -63,6 +64,35 @@ Some projects might define both build and runtime requirements that go beyond wh
 
 * If it is idiomatic in your language, prefer early handling of errors and return early (more info: [Avoid Else, Return Early](http://blog.timoxley.com/post/47041269194/avoid-else-return-early), [GuardClause](http://wiki.c2.com/?GuardClause) and [HandleErrorsInContext](http://wiki.c2.com/?HandleErrorsInContext)).
 
+## TODO comments
+
+The TODO comments should match the following format:
+
+```
+// KEYWORD(reference): comment text
+```
+
+A single line comment token must be used (`//` in case of C/C++, Go, etc). This is done to allow commenting large chunks of code with multi-line comment tokens (`/*`) during development and refactoring.
+
+The `KEYWORD` must be one of the following:
+
+- `TODO` - Something is missing or is not up to standard, but at the same time is not critical for the current implementation. It may be a missing optimization, a possible new feature, a refactoring opportunity or a deprecation note.
+- `FIXME` - There is a known issue or bug which needs to be addressed. Those comments should only be commited if a new bug was discovered during working on other feature/bugfix, if the fix requires significant design changes or you don't have knowledge to address it. In other cases it's better to try fixing the bug, of course.
+
+The `reference` must be either:
+
+- A Github issue reference: `TODO(#123)`. This is the preferred reference format, especially for `FIXME` comments. An issue will help track the technical debt associated with `TODO/FIXME` comments and will start a discussions about new features or design considerations to address those comments.
+- A Github username: `TODO(user)` - a temporary reference for minor improvements. This reference type is most useful for `TODO` comments: possible optimizations, refactorings, etc. The `user` is the author of the comment that can be contacted for the additional context. Make sure to write a descriptive comment text, so other team members can understand the reason without the need contacting the author for clarification.
+
+Multiline comments should have the same indentation (spaces, not tabs) based on the first letter on the comment:
```

Comment:
I'd rather not force this kind of indentation without a pretty good reason. I'd argue it's easier to read a comment that flows normally, e.g.,

```
// TODO(foo): Blah blah the thing is bad
// and the other things are bad too blah
```

That said: I think it's fine if someone _chooses_ to indent their comments this way. This is something where I think the guide should not take a strong position. At most maybe we should say "if you choose to indent the body of the comment, do so consistently".

Reply:
The indentation also helps to distinguish TODO comments from regular comments. For example:

```
// TODO: Special comment line 1.
//       Special comment line 2.
// Regular comment.
```

Instead of:

```
// TODO: Special comment line 1.
// Special comment line 2.
// Regular comment.
```

Reply:
For that example, I would recommend:

```
// TODO(ref): Special comment line 1,
// special comment line 2.
//
// Other comment as needed.
```

or to leave a blank between the two sections entirely. My main point is, since none of this is machine-checked, we should minimize the degree of detail the humans writing the comment and the reviewers need to remember.

Reply:
> none of this is machine-checked

It actually is. This format is highlighted and indexed by all JetBrains IDEs. Not sure how VScode does this though, so I'd also like to see an opinion of someone using it as well.

You may also argue that IDEs are not mandatory, and it's true, but the `TODO(user)` format is not machine-checked either, but we still want to make it standard for a reason.

Reply:
Just checked VScode. There is no built-in support for TODO comments and most extensions that I've seen doesn't seem to support multi-line comments at all.

The Go team uses a convention similar to one @creachadair mentioned.

So I'm changing the wording to use the convention proposed above.

Reply:
The new wording LGTM, thanks.

[the rest of this is optional rationale discussion]

> You may also argue that IDEs are not mandatory, and it's true, but the `TODO(user)` format is not machine-checked either, but we still want to make it standard for a reason.

Right, I agree it makes sense to have standards. I mean only that we should try to avoid making standards that add more or less friction depending on what development tools each person uses. 

As a principle, I believe we should have as few rules as possible about the formatting and styling of comments. I believe comments should be as low-friction as possible, emphasizing clear communication rather than appearance or machine-readability.

Simple advice like what you've written here is uncontroversial and helpful ("separate TODO comments from documentation"): It helps the reader distinguish which is which, and is easy to do. More detailed formatting and tagging rules (e.g., Javadoc, Doxygen, etc.) increase the burden of writing clear comments, and seem to reduce the quality of the content in favour of the presentation. 

Detailed rules also take up time in review, since either the reviewer has to notice a violation and ask for a correction, or some presubmit tool has to flag the problem and fail a check—either way, the author has to go back and push another round of commits for what is essentially cosmetic. If the rule is important enough that the extra steps are worthwhile, then that's probably OK. Otherwise, if  neither the reviewer nor a tool enforce the style, then usage will drift and we might as well not have a rule at all.